src/NHibernate/Transaction/AdoNetWithDistributedTransactionFactory.cs

```diff
@@ -29,7 +30,18 @@ public void EnlistInDistributedTransactionIfNeeded(ISessionImplementor session)
 			
 			if (System.Transactions.Transaction.Current == null)
 				return;
-			
+
+			var originatingSession = session.ConnectionManager.Session;
+			if (originatingSession != session)
+			{
+				session.TransactionContext = new DependentContext();
+			}
+
+			if (originatingSession.TransactionContext != null)
+				return;
+
+			session = originatingSession;
```

Comment:
Previous logic was causing each sharing session to enlist its own resource within the transaction. Then each was having its own prepare/transaction complete/... In non distributed case, that was working quite well, because it appears all those volatile resources are called in turn according to their enlistment order.

But with distributed transaction, they would have been called asynchronously, and then what a mess...

So I have changed the logic for having only the originating session enlisting itself.